File: Problem16.js
// time complexity: O(n)
// space complexity: O(n)
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    let pat = pattern.split("");
    let str = s.split(" ");
    let pMap = new Map();
    let sMap = new Map();
    
    if(pat.length !== str.length) return false;
    for(let i=0; i<pat.length; i++){
        if(pMap.has(pat[i])){
            if(pMap.get(pat[i]) !== str[i]){
                return false
            }
        } else {
            pMap.set(pat[i], str[i]);
        }
        
        if(sMap.has(str[i])){
            if(sMap.get(str[i]) !== pat[i]){
                return false
            }
        } else {
            sMap.set(str[i], pat[i]);
        }
        
    }
    return true;
};

File: problem 16.py
# Time Complexity : O(m+n)
# Space Complexity :O(m+n)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


# Your code here along with comments explaining your approach in three sentences only
# similar to isomorphic strings - map letter with the word in words separately and check

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()

        if len(words) != len(pattern):
            return False

        hm = {}
        hm1 = {}

        for i in range(len(pattern)):
            if (pattern[i] in hm and hm[pattern[i]] != words[i]) or (words[i] in hm1 and hm1[words[i]] != pattern[i]):
                return False
            hm[pattern[i]] = words[i]
            hm1[words[i]] = pattern[i]
        return True
        